src/ru/salauyou/utils/locks/LockKeeperV2.java

```diff
@@ -0,0 +1,305 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicIntegerArray;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+
+import ru.salauyou.locks.LockKeeper.LockType;
+import sun.misc.Unsafe;
+
+public class LockKeeperV2 {
+
+    /*
+     * The aims of this implementation:
+     * 1) getting rid of global lock for waiter queue
+     * 2) allow concurrent acquision of non-overlapping stripes
+     * 3) switch from Lock[] to int[]
+     * 
+     * TODO: make locks reentrant
+     * TODO: implement readlock
+     */
+    
+    final int mask;
+    final AtomicIntegerArray stripes;
+    final Queue<Waiter> waiters = new ConcurrentLinkedQueue<>();
+    final Map<Class<?>, Integer> shiftsForClasses;
+    
+    final static Unsafe U = UnsafeUtil.getUnsafe();
+    
+    
+    
+    /**
+     * Creates a new LockKeeper with specified number of stripes
+     * @param segmentSizeLog log₂ of stripe number for each class. 
+     *        E. g. "10" means "1024 stripes"
+     * @param classes classes for which separate segments (stripe sets)
+     *        should be created. Total number of stripes will be 
+     *        (|classes| + 1) * 2<sup>segmentSizeLog</sup>
+     */
+    public LockKeeperV2(int segmentSizeLog, Class<?>... classes) {
+        mask = (1 << segmentSizeLog) - 1;
+        int size = mask + 1;
+        if (classes.length > 0) {
+            shiftsForClasses = new HashMap<>();
+            for (Class<?> cl : classes)
+                shiftsForClasses.put(cl, (mask + 1) * (shiftsForClasses.size() + 1));
+            size *= shiftsForClasses.size() + 1;
+        } else 
+            shiftsForClasses = null;        
+        stripes = new AtomicIntegerArray(size);
+    }
+    
+    
+    
+    /**
+     * Creates a new LockKeeper with one segment of 1024 
+     * stripes for all classes
+     */
+    public LockKeeperV2() {
+        this(10);
+    }
+      
+    
+    public Lock lockAndGet(Object o) throws InterruptedException {
+        return lockAndGet(LockType.WRITE, new Object[]{o});
+    }
+    
+    
+    public Lock lockAndGet(LockType lockType, Object o) 
+                                              throws InterruptedException {
+        return lockAndGet(lockType, new Object[]{o});
+    }
+    
+    
+    public Lock lockAndGet(Object... objects) throws InterruptedException {
+        return lockAndGet(LockType.WRITE, objects);
+    }
+    
+    
+    
+    public Lock lockAndGet(LockType lockType, Object... objects) 
+                                              throws InterruptedException {
+        int[] locks = collectLocks(objects);
+        Thread t = tryGetLocks(locks);
+        if (t == null)
+            return new CompositeLock(locks, LockType.WRITE);
+        else {
+            Waiter w = new Waiter(t, locks);
+            waiters.add(w);
```

Comment:
race condition here: between `if (t == null)` and `waiters.add(w)` waiter w is invisible in queue traversal in `tryUnlockWaiters()`